lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,534 @@
+                "pbench-server", "token_expiration_duration"
+            )
+        except NoOptionError:
+            self.token_expire_duration = None
+
+    def post(self):
+        """
+        Post request for registering a new user.
+        This requires a JSON data with required user fields
+        {
+            "username": "username",
+            "password": "password",
+            "first_name": first_name,
+            "last_name": "last_name",
+            "email": "user@domain.com"
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        :return: JSON Payload
+        if we succeed to add a user entry in database, the returned response_object will look like following:
+            response_object = {
+                    "status": "success", # not present if the request fails
+                    "message": "Successfully registered."/"failure message",
+                }
+        To get the auth token user has to perform the login action
+        """
+        # get the post data
+        user_data = request.get_json()
+        if not user_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        username = user_data.get("username")
+        if not username:
+            self.logger.warning("Missing username field")
+            abort(
+                400, message="Missing username field",
+            )
+        username = username.lower()
+        if username == "admin":
+            self.logger.warning("User tried to register with admin username")
+            abort(
+                400, message="Please choose another username",
+            )
+
+        # check if provided username already exist
+        try:
+            user = User.query(username=user_data.get("username"))
+        except Exception:
+            self.logger.exception("Exception while querying user")
+            abort(500, message="INTERNAL ERROR")
+        if user:
+            self.logger.warning(
+                "A user tried to re-register. Username: {}", user.username
+            )
+            abort(403, message="Provided username is already in use.")
+
+        password = user_data.get("password")
+        if not password:
+            self.logger.warning("Missing password field")
+            abort(
+                400, message="Missing password field",
+            )
+
+        email_id = user_data.get("email")
+        if not email_id:
+            self.logger.warning("Missing email field")
+            abort(
+                400, message="Missing email field",
+            )
+        # check if provided email already exist
+        try:
+            user = User.query(email=user_data.get("email_id"))
+        except Exception:
+            self.logger.exception("Exception while querying user")
```

Comment:
Along with message `logger.exception` also tells us where the exception occurred so we dont need to explicitly say during user creation.

Reply:
True.  However, while it will tell you the function name and line number (I think), you would then have to check out the correct source module in order to be able to find the correct line (and, it might not be easy to figure out which commit that is); on the other hand, if you make the log message unique, then you can just search for it and you'll find the right line, even if the source that you are looking at has been modified.

Reply:
Fixed